File: backend/app/services/external_agent.py
"""Adapter for calling the internal chat / super-agent service.

This is the piece that lets the prompt-management system drive the *real* model
for flow-level RAGAS evaluation, instead of only running internally.

The external model reads its per-node SYSTEM_PROMPT + USER_PROMPT directly from
the active ``PM_NODE_PROMPT_VER`` row in the shared Oracle DB — so the chat
request itself only carries the user message + a user id. For A/B comparison,
``flow_service`` temporarily flips the active flag before invoking this adapter.

Contract:
    POST {EXTERNAL_AGENT_BASE_URL}
    headers:  {EXTERNAL_AUTH_HEADER|"auth-key"}: {EXTERNAL_AUTH_KEY}
              {EXTERNAL_USER_HEADER|"user-id"}:  {EXTERNAL_USER_ID}
    request:  {"message": "<test input>", "user_id": "pm-test"}
    response: {"response": "<answer>", "docs": [...], "urls": [...], ...}
The other response fields (service_id / session_id / user_id / trace_id / urls /
images / db_data / followup_questions / knowhows) are intentionally ignored.
"""
from __future__ import annotations

import json

import httpx

from app.core.config import get_settings

# Hardcoded session context sent as ``session_system_prompt``. The agent's field
# is a STRING, so this is a *stringified* JSON object (json.dumps keeps the
# embedded quotes valid); the agent json.loads it to read CUBE_CHANNEL_ID & co.
_SESSION_SYSTEM_PROMPT = json.dumps(
    {
        "CUBE_CHANNEL_ID": "509108549",
        "CUBE_USER_ID": "2074340",
        "CUBE_USER_NM": "김태윤",
        "TRACE_ID": "AI-20260416-171758-44399577",
    },
    ensure_ascii=False,
)


class ExternalAgentError(RuntimeError):
    """An external chat call failed or is misconfigured."""


def external_enabled() -> bool:
    """True when run_mode=external AND a base URL is configured."""
    s = get_settings()
    return s.run_mode.strip().lower() == "external" and bool(s.external_agent_base_url.strip())


def _base_url() -> str:
    url = get_settings().external_agent_base_url.strip().rstrip("/")
    if not url:
        raise ExternalAgentError("EXTERNAL_AGENT_BASE_URL is not set (.env)")
    return url


def _normalize_docs(raw: object) -> list[str]:
    """Coerce the response ``docs`` field into a list[str] (skip non-strings)."""
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for d in raw:
        if isinstance(d, str):
            out.append(d)
        elif isinstance(d, dict):
            # tolerate {"content": "..."} / {"text": "..."} shapes
            for k in ("content", "text", "body"):
                v = d.get(k)
                if isinstance(v, str) and v:
                    out.append(v)
                    break
    return out


def _parse_chat_response(resp: httpx.Response) -> dict:
    """Coerce the endpoint's reply into ``{response, docs, raw}``.

    The agent isn't consistent about its reply shape: sometimes a JSON object
    ``{"response": ..., "docs": [...]}``, sometimes a bare JSON string, and
    sometimes plain text (not JSON at all). ``resp.json()`` only handles the
    first; the bare-text case raises ``Expecting value: line 1 column 1``. So we
    parse defensively: a dict is read as the contract, anything else (str / list
    / number / non-JSON body) is treated as the answer text itself."""
    try:
        data = resp.json()
    except (json.JSONDecodeError, ValueError):
        # Not JSON at all — the whole body is the answer.
        text = resp.text.strip()
        return {"response": text, "docs": [], "raw": text}
    if isinstance(data, dict):
        return {
            "response": str(data.get("response") or ""),
            "docs": _normalize_docs(data.get("docs")),
            "raw": data,
        }
    # Valid JSON but not an object (e.g. a quoted string) — use it as the answer.
    return {"response": str(data), "docs": [], "raw": data}


def _request_headers(*, auth_key: str | None = None, user_id: str | None = None) -> dict[str, str]:
    """Auth + user-id headers. The header NAMES default to "auth-key" / "user-id"
    but are overridable via EXTERNAL_AUTH_HEADER / EXTERNAL_USER_HEADER (gateways
    differ — e.g. "Authorization" / "X-User-Id"). The header VALUES come from
    settings, but a non-None ``auth_key`` / ``user_id`` override them (used by the
    DB-free direct test, which lets the caller aim at an arbitrary endpoint).
    Empty values are dropped so an unconfigured header is simply omitted."""
    s = get_settings()
    auth_name = s.external_auth_header.strip() or "auth-key"
    user_name = s.external_user_header.strip() or "user-id"
    ak = (auth_key if auth_key is not None else s.external_auth_key).strip()
    uid = (user_id if user_id is not None else s.external_user_id).strip()
    headers: dict[str, str] = {}
    if ak:
        headers[auth_name] = ak
    if uid:
        headers[user_name] = uid
    return headers


def _chat_payload(*, message: str, user_id: str | None = None) -> dict:
    """Build the external chat-request body. The agent now expects more than just
    {message, user_id}: session_id / chat_type / a2a_remote_urls / is_super_agent
    / main_model_name / session_system_prompt. The extra fields come from the
    EXTERNAL_* settings (defaults mirror the agent's contract); empty
    main_model_name is sent as null. ``user_id`` overrides the configured one when
    given (direct test)."""
    s = get_settings()
    return {
        "message": message,
        "user_id": user_id if user_id is not None else s.external_user_id,
        "session_id": s.external_session_id,
        # The agent routes on chat_type; an empty value isn't a valid route, so
        # default to "default" when unset.
        "chat_type": "default",
        "a2a_remote_urls": None,
        "is_super_agent": s.external_is_super_agent,
        "main_model_name": s.external_main_model_name or None,
        "session_system_prompt": _SESSION_SYSTEM_PROMPT,
    }


async def run_flow(*, message: str, timeout_s: float = 60.0) -> dict:
    """POST one turn to the external chat endpoint and return ``{response, docs}``.

    The external model resolves its SYSTEM_PROMPT/USER_PROMPT/MODEL_NM from the
    active PM_NODE_PROMPT_VER row by itself — see module docstring.
    """
    payload = _chat_payload(message=message)
    headers = _request_headers()
    try:
        async with httpx.AsyncClient(timeout=timeout_s) as client:
            resp = await client.post(_base_url(), json=payload, headers=headers)
            resp.raise_for_status()
            parsed = _parse_chat_response(resp)
    except Exception as exc:  # noqa: BLE001
        raise ExternalAgentError(f"chat run failed: {exc}") from exc
    return {"response": parsed["response"], "docs": parsed["docs"]}


def ensure_direct_url(base_url: str | None = None) -> str:
    """Resolve the direct-call target URL (request override, else settings) and
    raise a clear error when none is configured. Returns the normalized URL."""
    url = (base_url or get_settings().external_agent_base_url).strip().rstrip("/")
    if not url:
        raise ExternalAgentError(
            "호출할 외부 API URL이 없습니다 — 요청에 base_url을 넣거나 EXTERNAL_AGENT_BASE_URL(.env)을 설정하세요"
        )
    return url


async def run_direct(
    *,
    message: str,
    base_url: str | None = None,
    auth_key: str | None = None,
    user_id: str | None = None,
    timeout_s: float = 60.0,
) -> dict:
    """One-shot direct call to the external chat endpoint — NO DB, NO dataset,
    NO scoring. Sends ``message`` straight to the endpoint and returns its answer
    as-is (parsed ``response`` + ``docs`` plus the full ``raw`` body).

    Unlike :func:`run_flow` this does not require RUN_MODE=external and lets the
    caller override the target URL / auth / user-id, so an arbitrary endpoint can
    be smoke-tested without any .env or DB setup. The external model still resolves
    its own prompt on its side — this system just relays the message and answer.
    """
    url = ensure_direct_url(base_url)
    payload = _chat_payload(message=message, user_id=user_id)
    headers = _request_headers(auth_key=auth_key, user_id=user_id)
    # TEMP debug: print the exact JSON body we send so we can confirm
    # session_system_prompt is actually included in the outgoing request.
    # (print, not logger — the app configures no logging handler, so INFO is dropped.)
    print(f"[DIRECT] → {url} body={json.dumps(payload, ensure_ascii=False)}", flush=True)
    try:
        async with httpx.AsyncClient(timeout=timeout_s) as client:
            resp = await client.post(url, json=payload, headers=headers)
            resp.raise_for_status()
            # TEMP debug: surface what actually came back so a non-JSON body
            # (SSE stream / HTML error page / empty) is visible instead of being
            # hidden behind a bare "Expecting value: line 1 column 1" decode error.
            ctype = resp.headers.get("content-type", "")
            print(
                f"[DIRECT] ← status={resp.status_code} content-type={ctype!r} "
                f"len={len(resp.content)} body[:500]={resp.text[:500]!r}",
                flush=True,
            )
            return _parse_chat_response(resp)
    except ExternalAgentError:
        raise
    except Exception as exc:  # noqa: BLE001
        raise ExternalAgentError(f"direct call failed: {exc}") from exc


async def stub_run_flow(*, message: str, **_: object) -> dict:
    """TEMPORARY in-process stand-in for ``run_flow`` (no external endpoint yet).

    Returns a deterministic placeholder answer in the same shape as ``run_flow``
    so flow-level RAGAS runs end-to-end while the real chat/super-agent is not
    connected. Swap by setting ``RUN_MODE=external`` + ``EXTERNAL_AGENT_BASE_URL``.
    """
    return {"response": f"[stub answer] {message}".strip(), "docs": []}


async def flow_answer(*, message: str) -> dict:
    """One flow answer: the real chat endpoint when external is enabled, else the stub."""
    if external_enabled():
        return await run_flow(message=message)
    return await stub_run_flow(message=message)

File: src/lib/services/externalAgent.ts
import { getAgentConfig, getCallTimeoutMs, getFlowBaseUrl, getFlowHeaders } from "@/lib/config";
import type { EndpointHeader } from "@/lib/types";
import { ApiError, badGateway, fetchWithTimeout } from "@/lib/http";

// Session context sent as ``session_system_prompt`` (a STRING that is a stringified
// JSON object — the agent json.loads it to read CUBE_CHANNEL_ID & co). Channel and
// user name are fixed for this tool; the user id is whoever is calling (config
// agent.userId / the request's user_id) and TRACE_ID is issued per call.
const CUBE_CHANNEL_ID = "11111111";
const CUBE_CHANNEL_NM = "프롬프트 자동화 테스트";
const CUBE_USER_NM = "이억수";

let traceDay = "";
let traceSeq = 0;

/** Seed the counter from the wall clock (centiseconds since midnight) so a
 * restart cannot reissue ids an earlier process already used. That matters
 * because the agent stores captured variables in PTX_TRACE_HIS under this id: a
 * reused id would let a stale row be read back as this call's value. */
function seedSeq(d: Date): number {
  return (d.getHours() * 3600 + d.getMinutes() * 60 + d.getSeconds()) * 100;
}

/** ``PTX-YYYYMMDD-NNNN`` — monotonic within a day, across restarts.
 * Exported because a caller that has to stage something for the agent under this
 * id (PTX_CALL_MAS) needs the id *before* the request goes out. */
export function nextTraceId(): string {
  const d = new Date();
  const day =
    `${d.getFullYear()}` +
    `${String(d.getMonth() + 1).padStart(2, "0")}` +
    `${String(d.getDate()).padStart(2, "0")}`;
  if (day !== traceDay) {
    traceDay = day;
    traceSeq = seedSeq(d);
  }
  traceSeq += 1;
  return `PTX-${day}-${String(traceSeq).padStart(4, "0")}`;
}

function sessionContext(userId: string, traceId: string): Record<string, string> {
  return {
    CUBE_CHANNEL_ID,
    CUBE_CHANNEL_NM,
    CUBE_USER_ID: userId,
    CUBE_USER_NM,
    TRACE_ID: traceId,
  };
}

export interface AgentAnswer {
  response: string;
  docs: string[];
  raw?: Record<string, unknown> | unknown[] | string;
  /** TRACE_ID sent with this call — the key the agent writes PTX_TRACE_HIS under. */
  traceId?: string;
  /** Request sent → first token of the answer arrived, in ms. Only a streaming
   * (text/event-stream) reply has one: it is the part of the wait that happens
   * before any of the answer is written, so it holds queue time and drops the
   * generation time that makes the total depend on how long the answer got.
   * null when the endpoint answered in a single body. */
  ttftMs?: number | null;
}

/** A/B side of a run — each side may have its own configured endpoint. */
export type FlowSide = "a" | "b";

export function externalEnabled(): boolean {
  const a = getAgentConfig();
  return a.runMode === "external" && (a.a.url || a.b.url).length > 0;
}

function normalizeDocs(raw: unknown): string[] {
  if (!Array.isArray(raw)) return [];
  const out: string[] = [];
  for (const d of raw) {
    if (typeof d === "string") out.push(d);
    else if (d && typeof d === "object") {
      for (const k of ["content", "text", "body"]) {
        const v = (d as Record<string, unknown>)[k];
        if (typeof v === "string" && v) {
          out.push(v);
          break;
        }
      }
    }
  }
  return out;
}

function collectTxt(obj: unknown, out: string[]): void {
  if (Array.isArray(obj)) {
    for (const item of obj) collectTxt(item, out);
  } else if (obj && typeof obj === "object") {
    const o = obj as Record<string, unknown>;
    if (o.type === "txt" && typeof o.value === "string") out.push(o.value);
  }
}

/** The answer text carried by ONE `data:` line — "" when the line carries none
 * (a keep-alive, `[DONE]`, or a frame whose payload is metadata rather than
 * text). Shared so the incremental read and the final parse agree on what
 * counts as answer text; TTFT is defined as the first line this returns
 * something for. */
function sseLineText(line: string): string {
  const trimmed = line.trim();
  if (!trimmed.startsWith("data:")) return "";
  const payload = trimmed.slice("data:".length).trim();
  if (!payload || payload === "[DONE]") return "";
  let obj: unknown;
  try {
    obj = JSON.parse(payload);
  } catch {
    try {
      obj = JSON.parse(`[${payload}]`);
    } catch {
      return "";
    }
  }
  const parts: string[] = [];
  collectTxt(obj, parts);
  return parts.join("");
}

/** Aggregate a text/event-stream reply into {response, docs, raw}. */
function parseSse(text: string): AgentAnswer {
  const parts: string[] = [];
  for (const line of text.split(/\r?\n/)) {
    const t = sseLineText(line);
    if (t !== "") parts.push(t);
  }
  return { response: parts.join(""), docs: [], raw: text };
}

/**
 * The response body, plus when its first token of answer text arrived.
 *
 * `resp.text()` would be shorter, but it resolves only once the last byte is
 * in — and by then the moment the answer *started* is gone. Reading the stream
 * chunk by chunk costs nothing extra and recovers TTFT: request sent → first
 * token, which is queue wait + prefill without any of the generation time that
 * dominates the total. The text handed back is byte-for-byte what `text()`
 * would have produced, so every parser below is unaffected.
 *
 * `ttftMs` is null when the endpoint does not stream — one JSON body has no
 * first token, and its arrival time IS its completion time.
 */
async function readBodyTimed(resp: Response, startedAt: number, sse: boolean): Promise<{ text: string; ttftMs: number | null }> {
  if (!resp.body) return { text: await resp.text(), ttftMs: null };
  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let text = "";
  let pending = ""; // the tail of the last chunk: possibly half a frame
  let ttftMs: number | null = null;
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      const chunk = decoder.decode(value, { stream: true });
      text += chunk;
      if (!sse || ttftMs !== null) continue;
      pending += chunk;
      const lines = pending.split(/\r?\n/);
      // A frame torn across two TCP writes is not a token yet; its tail waits
      // for the rest rather than being parsed as truncated JSON.
      pending = lines.pop() ?? "";
      for (const line of lines) {
        if (sseLineText(line) !== "") {
          ttftMs = Date.now() - startedAt;
          break;
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
  text += decoder.decode();
  return { text, ttftMs };
}

async function parseChatResponse(resp: Response, startedAt: number): Promise<AgentAnswer> {
  const ctype = (resp.headers.get("content-type") ?? "").toLowerCase();
  const isSse = ctype.includes("text/event-stream");
  const { text, ttftMs } = await readBodyTimed(resp, startedAt, isSse);
  if (isSse) return { ...parseSse(text), ttftMs };
  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch {
    const t = text.trim();
    return { response: t, docs: [], raw: t };
  }
  if (data && typeof data === "object" && !Array.isArray(data)) {
    let o = data as Record<string, unknown>;
    // A 200 carrying an `error` object is still a failure — raise it rather than
    // scoring the error body as if it were an answer. The endpoint's own code and
    // message are the whole content of the failure, so both are surfaced.
    if (o.error && typeof o.error === "object") {
      const e = o.error as Record<string, unknown>;
      const code = String(e.code ?? "").trim();
      // `data: null` carries nothing; appending it just puts a bare "null" in
      // front of the user.
      const detail =
        e.data === undefined || e.data === null ? "" : ` ${JSON.stringify(e.data)}`;
      throw new Error(
        `API 오류${code ? ` ${code}` : ""} — ${String(e.message ?? "")}${detail}`,
      );
    }
    // Some endpoints wrap the payload in `result`; score what is inside it.
    const result = o.result;
    if (result !== undefined) {
      if (result && typeof result === "object" && !Array.isArray(result)) {
        o = result as Record<string, unknown>;
      } else {
        const txt = typeof result === "string" ? result : JSON.stringify(result);
        return { response: txt, docs: [], raw: Array.isArray(result) ? result : txt };
      }
    }
    // Endpoints that reply with a different envelope (no `response` key, e.g.
    // {header, body}) keep their whole JSON as the answer so 정답 일치 can
    // compare it instead of scoring an empty string.
    return {
      response: o.response !== undefined ? String(o.response) : JSON.stringify(o),
      docs: normalizeDocs(o.docs),
      raw: o,
    };
  }
  return { response: String(data), docs: [], raw: data as string };
}

/** Content-Type plus the headers of whatever endpoint answers this call.
 * ``registered`` are the headers saved with the endpoint the run picked in the
 * settings registry; without it the side's config headers are used, so a run
 * that names no endpoint behaves exactly as before. ``authKey`` is the key typed
 * into the UI for a one-off call; it overrides the FIRST header's value, that
 * slot being the endpoint's credential. */
function requestHeaders(
  side?: FlowSide | null,
  authKey?: string | null,
  registered?: EndpointHeader[] | null,
): Record<string, string> {
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  const configured = registered?.length ? registered : getFlowHeaders(side);
  for (const h of configured) headers[h.name] = h.value;
  const override = (authKey ?? "").trim();
  if (override && configured.length) headers[configured[0].name] = override;
  return headers;
}

/** The request body — exactly these five keys, the endpoint's spec and nothing
 * beyond it — plus the TRACE_ID embedded in it, which the caller needs to look
 * up whatever the agent captured mid-flow. */
function buildPayload(
  message: string,
  userId?: string | null,
  traceId?: string | null,
): { body: Record<string, unknown>; traceId: string } {
  const uid = userId ?? getAgentConfig().userId;
  // A caller that already staged rows under an id passes it in; everyone else
  // gets a fresh one.
  const tid = traceId || nextTraceId();
  return {
    body: {
      message,
      user_id: uid,
      session_id: "",
      chat_type: "default",
      // A STRING that is a stringified JSON object — the agent json.loads it.
      session_system_prompt: JSON.stringify(sessionContext(uid, tid)),
    },
    traceId: tid,
  };
}

/** Read per call rather than captured at module load: the config is reloadable
 * and a run should honour the value in effect when it starts. */
function callTimeoutMs(): number {
  return getCallTimeoutMs();
}

async function post(
  url: string,
  body: unknown,
  headers: Record<string, string>,
  timeoutMs = callTimeoutMs(),
): Promise<Response> {
  // fetchWithTimeout, not fetch: it puts connect/headers/body under this same
  // deadline. Plain fetch would give up on the connect after Node's own 10s and
  // report a network error long before the configured limit.
  return fetchWithTimeout(url, { method: "POST", headers, body: JSON.stringify(body) }, timeoutMs);
}

function oneLine(s: string): string {
  return s.replace(/\s+/g, " ").trim();
}

/** A non-2xx reply, with the body — the endpoint's own explanation of the
 * refusal ("field required", a stack trace, a gateway notice) is the whole
 * point; `HTTP 500` on its own tells nobody anything. */
async function httpError(resp: Response, body: unknown): Promise<Error> {
  const text = oneLine(await resp.text().catch(() => ""));
  const status = `HTTP ${resp.status}${resp.statusText ? ` ${resp.statusText}` : ""}`;
  // A rejected request shape is the usual cause of 400/422, so those — and only
  // those — are worth telling the user what we actually sent.
  const sent = resp.status === 400 || resp.status === 422 ? sentTag(body) : "";
  return new Error(`${status}${text ? ` — ${text.slice(0, 900)}` : ""}${sent}`);
}

/** Short "what we actually sent" tag appended to failures — the terminal log is
 * not always visible, so the shape of the request has to reach the UI. */
function sentTag(body: unknown): string {
  if (!body || typeof body !== "object") return "";
  return ` | sent: keys=[${Object.keys(body as Record<string, unknown>).join(",")}]`;
}

/** `fetch` only takes absolute URLs; a host written without a scheme fails as
 * "Failed to parse URL from …", naming the path rather than the setting. */
function requireScheme(url: string, where: string): string {
  if (!/^https?:\/\//i.test(url)) {
    throw badGateway(`${where} 에 http:// 또는 https:// 가 없습니다 (현재: "${url}")`);
  }
  return url;
}

/** Which config key a side's URL lives under — for error messages only. */
function urlSetting(side?: FlowSide | null): string {
  return `agent.${side === "b" ? "b" : "a"}.url`;
}

function baseUrl(side?: FlowSide | null): string {
  const setting = urlSetting(side);
  const url = getFlowBaseUrl(side).trim().replace(/\/+$/, "");
  if (!url) throw badGateway(`${setting} 이 설정되어 있지 않습니다 (config.yml)`);
  return requireScheme(url, setting);
}

/** ``side`` picks which configured endpoint answers when nothing was typed —
 * a manual A/B compares agent.a.url against agent.b.url. */
export function ensureDirectUrl(override?: string | null, side: FlowSide = "a"): string {
  const url = (override || getFlowBaseUrl(side)).trim().replace(/\/+$/, "");
  const setting = urlSetting(side);
  if (!url) {
    throw badGateway(
      `호출할 외부 API URL이 없습니다 — 요청에 base_url을 넣거나 config.yml 의 ${setting} 을 설정하세요`,
    );
  }
  return requireScheme(url, override ? "입력한 Base URL" : setting);
}

/** POST one turn to the external chat endpoint (RUN_MODE=external).
 * ``urlOverride`` pins this call to a specific endpoint; otherwise the side's
 * configured endpoint is used (agent.a.url / agent.b.url). */
export async function runFlow(
  message: string,
  urlOverride?: string | null,
  side?: FlowSide | null,
  traceIdIn?: string | null,
  headers?: EndpointHeader[] | null,
): Promise<AgentAnswer> {
  const url = urlOverride ? ensureDirectUrl(urlOverride) : baseUrl(side);
  const { body, traceId } = buildPayload(message, null, traceIdIn);
  // Started here, immediately before the request goes out — URL building and
  // payload assembly above are ours, and TTFT must not carry them.
  const startedAt = Date.now();
  const resp = await post(url, body, requestHeaders(side, null, headers));
  if (!resp.ok) throw await httpError(resp, body);
  const parsed = await parseChatResponse(resp, startedAt);
  return { response: parsed.response, docs: parsed.docs, traceId, ttftMs: parsed.ttftMs ?? null };
}

/** One-shot direct call — no DB, no scoring; caller may override URL/auth/user. */
export async function runDirect(args: {
  message: string;
  baseUrl?: string | null;
  authKey?: string | null;
  userId?: string | null;
  /** Which configured endpoint answers when no URL is typed. Only a manual A/B
   * passes 'b'; every other direct call is side A. */
  side?: FlowSide | null;
  /** Headers of the registered endpoint this call picked, when it picked one. */
  headers?: EndpointHeader[] | null;
  /** Pre-issued correlation id when the caller staged rows under it. */
  traceId?: string | null;
}): Promise<AgentAnswer> {
  const side = args.side ?? "a";
  const url = ensureDirectUrl(args.baseUrl, side);
  const { body, traceId } = buildPayload(args.message, args.userId, args.traceId);
  const startedAt = Date.now();
  try {
    const resp = await post(url, body, requestHeaders(side, args.authKey, args.headers));
    if (!resp.ok) throw await httpError(resp, body);
    return { ...(await parseChatResponse(resp, startedAt)), traceId };
  } catch (e) {
    // Carriage, not handling: a raw Error reaching a route handler comes out of
    // `errorResponse` as "internal server error", which throws the endpoint's
    // own words away. Re-thrown with the message untouched so the screen shows
    // what the other side actually said. A dataset run needs none of this — its
    // per-case catch already records the same text.
    if (e instanceof ApiError) throw e;
    throw badGateway(e instanceof Error ? e.message || String(e) : String(e));
  }
}

/** Deterministic in-process stand-in for runFlow (no external endpoint). */
export function stubRunFlow(message: string): AgentAnswer {
  return { response: `[stub answer] ${message}`.trim(), docs: [] };
}

/** One flow answer. A URL typed into the UI always wins (and is called even in
 * stub mode); otherwise the side's configured endpoint is used when external is
 * enabled, else the stub. */
export async function flowAnswer(
  message: string,
  urlOverride?: string | null,
  side?: FlowSide | null,
  traceId?: string | null,
  headers?: EndpointHeader[] | null,
): Promise<AgentAnswer> {
  // A picked endpoint brings its own headers; without one the side's config
  // headers decide, even when the URL came from the UI.
  if (urlOverride) return runFlow(message, urlOverride, side, traceId, headers);
  if (externalEnabled()) return runFlow(message, null, side, traceId, headers);
  return stubRunFlow(message);
}
